File: sdk/typescript/src/modules/activity.ts
import type { ChainClient } from "../infrastructure/chain-client.js";
import type { SigningService } from "../infrastructure/signing-service.js";
import type { ResolvedTxConfig } from "../types/config.js";
import type {
  SubmitActivityResponse,
  GetActivitiesResponse,
  ActivityItem,
  GetQuotaResponse,
} from "../types/responses.js";
import { MsgSubmitActivity } from "../infrastructure/messages.js";
import { ChainClientAdapter, executeTx } from "../infrastructure/tx-pipeline.js";
import type { PipelineConfig } from "../infrastructure/tx-pipeline.js";
import { isValidActivityHash, isValidContentUri } from "../utils/hash.js";
import { computeEpoch, computeWindow } from "../utils/epoch.js";
import { ValidationError, TransactionError } from "../errors/errors.js";
import {
  ERR_INVALID_ACTIVITY_HASH,
  ERR_INVALID_CONTENT_URI,
  ERR_BROADCAST_FAILED,
} from "../constants/errors.js";

export class ActivityModule {
  constructor(
    private readonly chainClient: ChainClient,
    private readonly signingService: SigningService,
    private readonly txConfig: ResolvedTxConfig,
  ) {}

  async submit(
    activityHash: string,
    contentUri: string,
  ): Promise<SubmitActivityResponse> {
    if (!isValidActivityHash(activityHash)) {
      throw new ValidationError(
        "activity hash must be exactly 64 hex characters (32 bytes)",
        ERR_INVALID_ACTIVITY_HASH,
      );
    }
    if (!isValidContentUri(contentUri)) {
      throw new ValidationError(
        "content URI must not be empty",
        ERR_INVALID_CONTENT_URI,
      );
    }

    const submitter = await this.signingService.getAddress();
    const msg = new MsgSubmitActivity(submitter, activityHash, contentUri);

    const querier = new ChainClientAdapter(this.chainClient);
    const pipelineCfg: PipelineConfig = {
      chainId: this.txConfig.chain_id,
      gasPrice: this.txConfig.gas_price,
      confirmTimeoutMs: this.txConfig.confirm_timeout_ms,
      pollIntervalMs: this.txConfig.confirm_poll_interval_ms,
    };

    const result = await executeTx(querier, this.signingService, pipelineCfg, {
      message: msg,
    });

    if (result.code !== 0) {
      throw new TransactionError(
        `activity submit failed with code ${result.code}: ${result.rawLog}`,
        ERR_BROADCAST_FAILED,
      );
    }

    // Compute epoch/window from block height
    const params = await this.getActivityParams();
    const epoch = computeEpoch(
      result.height,
      params.epoch_length,
    );
    const window = computeWindow(
      result.height,
      params.epoch_length,
      params.windows_per_epoch,
    );

    // Query remaining quota
    let quotaRemaining = 0;
    try {
      const quota = await this.getQuota();
      quotaRemaining = quota.quota_remaining;
    } catch {
      // Quota query failure is non-fatal
    }

    return {
      tx_hash: result.txHash,
      block_height: result.height,
      epoch_number: epoch,
      window_number: window,
      quota_remaining: quotaRemaining,
    };
  }

  async getActivities(
    nodeId?: string,
    epochNumber?: number,
  ): Promise<GetActivitiesResponse> {
    const effectiveNodeId = nodeId ?? (await this.resolveOwnNodeId());
    const effectiveEpoch = epochNumber ?? await this.computeCurrentEpoch();

    const response = await this.chainClient.queryRest<{
      activities: Array<{
        activity_hash: string;
        content_uri: string;
        block_height: string;
      }>;
    }>(
      `/seocheon/activity/v1/nodes/${effectiveNodeId}/activities?epoch=${effectiveEpoch}`,
    );

    const params = await this.getActivityParams();
    const activities: ActivityItem[] = (response.activities ?? []).map(
      (record) => {
        const blockHeight = parseInt(record.block_height, 10);
        return {
          activity_hash: record.activity_hash,
          content_uri: record.content_uri,
          block_height: blockHeight,
          window_number: computeWindow(
            blockHeight,
            params.epoch_length,
            params.windows_per_epoch,
          ),
          tx_hash: "",
        };
      },
    );

    return {
      activities,
      total_count: activities.length,
    };
  }

  async getQuota(): Promise<GetQuotaResponse> {
    const nodeId = await this.resolveOwnNodeId();
    const epochNumber = await this.computeCurrentEpoch();

    const response = await this.chainClient.queryRest<{
      summary: { active_windows: string; total_activities: string; eligible: boolean };
      quota_used: string;
      quota_limit: string;
    }>(`/seocheon/activity/v1/nodes/${nodeId}/epochs/${epochNumber}`);

    const quotaUsed = parseInt(response.quota_used, 10);
    const quotaLimit = parseInt(response.quota_limit, 10);

    return {
      epoch_number: epochNumber,
      quota_total: quotaLimit,
      quota_used: quotaUsed,
      quota_remaining: quotaLimit - quotaUsed,
      is_feegrant: false,
      feegrant_expiry: null,
    };
  }

  private async resolveOwnNodeId(): Promise<string> {
    const address = await this.signingService.getAddress();
    const response = await this.chainClient.queryRest<{
      node: { id: string };
    }>(`/seocheon/node/v1/nodes/by-agent/${address}`);
    return response.node.id;
  }

  private async computeCurrentEpoch(): Promise<number> {
    try {
      const block = await this.chainClient.getLatestBlock();
      const params = await this.getActivityParams();
      return computeEpoch(block.header.height, params.epoch_length);
    } catch {
      return 0;
    }
  }

  private async getActivityParams(): Promise<{
    epoch_length: number;
    windows_per_epoch: number;
    min_active_windows: number;
  }> {
    const response = await this.chainClient.queryRest<{
      params: {
        epoch_length: string;
        windows_per_epoch: string;
        min_active_windows: string;
      };
    }>("/seocheon/activity/v1/params");
    return {
      epoch_length: parseInt(response.params.epoch_length, 10),
      windows_per_epoch: parseInt(response.params.windows_per_epoch, 10),
      min_active_windows: parseInt(response.params.min_active_windows, 10),
    };
  }
}

File: sdk/python/seocheon/modules/activity.py
"""Activity module for the Seocheon SDK."""

from __future__ import annotations

from seocheon.constants.chain import EPOCH_LENGTH, WINDOWS_PER_EPOCH
from seocheon.errors.errors import (
    ErrInvalidActivityHash,
    ErrInvalidContentURI,
    ErrQueryFailed,
    ErrSubmitterNotRegistered,
    abci_code_to_error,
)
from seocheon.infrastructure.chain_client import ChainClient
from seocheon.infrastructure.signing.service import SigningService
from seocheon.internal.tx.chain_adapter import ChainClientAdapter
from seocheon.internal.tx.messages import MsgSubmitActivity
from seocheon.internal.tx.pipeline import PipelineConfig, default_pipeline_config, execute_tx
from seocheon.internal.tx.types import TxRequest
from seocheon.types.responses import (
    ActivityItem,
    GetActivitiesResponse,
    GetQuotaResponse,
    SubmitActivityResponse,
)
from seocheon.utils.epoch import compute_epoch, compute_window
from seocheon.utils.hash import verify_activity_hash


class ActivityModule:
    """Provides activity-related operations."""

    def __init__(self, client: ChainClient, signer: SigningService, chain_id: str) -> None:
        self._client = client
        self._signer = signer
        self._tx_querier = ChainClientAdapter(client)
        self._tx_config: PipelineConfig = default_pipeline_config(chain_id)

    async def submit(self, activity_hash: str, content_uri: str) -> SubmitActivityResponse:
        """Submit an activity hash to the chain."""
        if not verify_activity_hash(activity_hash):
            raise ErrInvalidActivityHash
        if not content_uri:
            raise ErrInvalidContentURI

        msg = MsgSubmitActivity(
            submitter=self._signer.get_address(),
            activity_hash=activity_hash,
            content_uri=content_uri,
        )

        result = await execute_tx(self._tx_querier, self._signer, self._tx_config, TxRequest(message=msg))

        if result.code != 0:
            raise abci_code_to_error(result.code)

        params = await self._get_activity_params()
        epoch_number = compute_epoch(result.height, params["epoch_length"])
        window_number = compute_window(result.height, params["epoch_length"], params["windows_per_epoch"])

        quota_remaining = 0
        try:
            quota = await self.get_quota()
            quota_remaining = quota.quota_remaining
        except Exception:
            pass

        return SubmitActivityResponse(
            tx_hash=result.tx_hash,
            block_height=result.height,
            epoch_number=epoch_number,
            window_number=window_number,
            quota_remaining=quota_remaining,
        )

    async def get_activities(self, node_id: str = "", epoch_number: int = 0) -> GetActivitiesResponse:
        """Return activity submission history for a node."""
        effective_node_id = node_id or await self._resolve_own_node_id()
        effective_epoch = epoch_number or await self._compute_current_epoch()

        path = f"/seocheon/activity/v1/nodes/{effective_node_id}/activities?epoch={effective_epoch}"
        try:
            data = await self._client.query_rest(path)
        except Exception as e:
            raise ErrQueryFailed from e

        params = await self._get_activity_params()
        items = []
        for a in data.get("activities", []):
            height = int(a.get("block_height", 0))
            wn = compute_window(height, params["epoch_length"], params["windows_per_epoch"])
            items.append(ActivityItem(
                activity_hash=a.get("activity_hash", ""),
                content_uri=a.get("content_uri", ""),
                block_height=height,
                window_number=wn,
            ))

        return GetActivitiesResponse(activities=items, total_count=len(items))

    async def get_quota(self) -> GetQuotaResponse:
        """Return the remaining activity submission quota for the current epoch."""
        node_id = await self._resolve_own_node_id()
        epoch_num = await self._compute_current_epoch()

        path = f"/seocheon/activity/v1/nodes/{node_id}/epochs/{epoch_num}"
        try:
            data = await self._client.query_rest(path)
        except Exception as e:
            raise ErrQueryFailed from e

        used = int(data.get("quota_used", 0))
        limit = int(data.get("quota_limit", 0))

        agent_addr = self._signer.get_address()
        is_feegrant = await self._check_feegrant(agent_addr)

        return GetQuotaResponse(
            epoch_number=epoch_num,
            quota_total=limit,
            quota_used=used,
            quota_remaining=limit - used,
            is_feegrant=is_feegrant,
        )

    async def _resolve_own_node_id(self) -> str:
        agent_addr = self._signer.get_address()
        path = f"/seocheon/node/v1/nodes/by-agent/{agent_addr}"
        try:
            data = await self._client.query_rest(path)
        except Exception as e:
            raise ErrSubmitterNotRegistered from e
        node_id = data.get("node", {}).get("id", "")
        if not node_id:
            raise ErrSubmitterNotRegistered
        return node_id

    async def _get_activity_params(self) -> dict:
        try:
            data = await self._client.query_rest("/seocheon/activity/v1/params")
            params = data.get("params", {})
            el = int(params.get("epoch_length", 0)) or EPOCH_LENGTH
            wpe = int(params.get("windows_per_epoch", 0)) or WINDOWS_PER_EPOCH
            return {"epoch_length": el, "windows_per_epoch": wpe}
        except Exception:
            return {"epoch_length": EPOCH_LENGTH, "windows_per_epoch": WINDOWS_PER_EPOCH}

    async def _compute_current_epoch(self) -> int:
        block = await self._client.get_latest_block()
        params = await self._get_activity_params()
        return compute_epoch(block["height"], params["epoch_length"])

    async def _check_feegrant(self, agent_addr: str) -> bool:
        try:
            path = f"/cosmos/feegrant/v1beta1/allowance/feegrant_pool/{agent_addr}"
            data = await self._client.query_rest(path)
            return data.get("allowance") is not None
        except Exception:
            return False
